### PR TITLE
ParaView improvements [paraview-improvements]

### LINKED ISSUE #1284: ParaViewDataCollection does not respect SetPrefixPath appropriately
Using `SetPrefixPath` only saves the data in the prefix, but not the `pvd` file.

### PR DESCRIPTION
**Pull request by pazner**
This PR adds several new features and makes some improvements to the `ParaViewDataCollection` class:

- High-order output for all element types: tensor products, simplices, and prisms
- Option to output in binary format (either Float64 or Float32)
- Compressed binary output if compiled with gzstream support
- Write files in a way that allows ParaView to read “partially written” data collections (i.e. during time stepping, before completion)
- Allow for output of high-order VTU format meshes
- Code reorganization: avoid duplicating pre-existing functionality from the `DataCollection` base class
- Bug fix: respect precision option for ASCII output
- Bug fix: respect prefix path for data collection (fixes #1284)

Related to #678 and GLVis/glvis#64.

An example of the high-order visualization:

tri

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1296](https://github.com/mfem/mfem/pull/1296) | @pazner | @tzanio | @bslazarov + @jandrej | 02/16/20 | 02/24/20 | 02/27/20 | |

### CODE REVIEW

**Comment by pazner** [comment]
The CI checks all look good except for the large file modification check because of additions and changes to `mesh.cpp`. Some of these changes are inevitable, but I also added a lot of VTK-specific code, which might be better-off being moved to a different file (e.g. creating `vtk.cpp` or similar). I know @camierjs had some thoughts about this.

I would appreciate feedback about the best way to proceed.

**Comment by tzanio** [comment]
I'm totally OK with creating `vtk.?pp` for  VTK-specific code. In principle, splitting code in a new file is totally fine as long as: 1) it is more than just a few lines; and 2) does not split class implementation on multiple files. There could be exception to both of these too.

**Comment by tzanio** [comment]
Please update `CHANGELOG`

**Comment by pazner** [comment]
I updated `CHANGELOG` and moved a lot of the VTK-specific logic (node numbering, binary headers, encoding, etc.) to the new file `vtk.cpp`.

File: fem/datacollection.hpp [hunk 1]
```diff
@@ -530,15 +513,33 @@ class ParaViewDataCollection : public DataCollection
...
+   /// 9, 1 being the fastest, and 9 being the best compression. Compression
+   /// only takes effect if the output format is BINARY or BINARY32. MFEM must
+   /// be compiled with MFEM_USE_GZSTREAM = YES.
+   void SetCompressionLevel(int compression_level_);
 
-#ifdef MFEM_USE_MPI
-   static int create_directory(const std::string &dir_name, int myid,
-                               MPI_Comm mycom);
-#endif
+   /// Enable or diable zlib compression. If the input is true, use the default
```
Comment (jandrej) [inline fem/datacollection.hpp]: s/diable/disable

**Comment by pazner** [comment]
> LGTM. I'd like to see the valgrind test results but from the implementation point this is good.

No issues with valgrind running ex5 or ex9 with ParaView output: `All heap blocks were freed -- no leaks are possible`

File: general/binaryio.hpp [hunk 2]
```diff
@@ -38,6 +39,18 @@ inline T read(std::istream& is)
    return value;
 }
 
+template <typename T>
+void AppendBytes(std::vector<char> &vec, const T &val)
+{
+   for(char ib=0; ib<sizeof(T); ++ib)
```
Comment (bslazarov) [inline general/binaryio.hpp]: ib = 0; ??formating

**Comment by pazner** [comment]
Good point about the endianness. Now I detect the endianness and output the proper `byte_order` string. It shouldn't really be an issue because I think even the POWER machines are running in little endian mode...

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by rcarson3** [comment]
While testing this with ExaConstit, I ran into an issue where my GridFunctions that make use of an L2 finite element space are being outputted to 0 whenever I run my app using a CUDA runtime. I don't have this issue with the ConduitDataCollection or when I set my runtime to being the CPU.

It does appear though that this bug existed with the original Paraview implementation as well.

**Comment by v-dobrev** [comment]
@rcarson3, can you try building MFEM in debug mode? That should catch data use on the host if the data is actually on the device. Hopefully that will tell us where we need to look to fix the issue.

**Comment by rcarson3** [comment]
Hey @v-dobrev I just ran it with the debug assertion and there was definitely an invalid host pointer assertion being thrown. So, it appears it was outside of ParaviewDataCollection call at least initially. 

It appears that it's being thrown in ```ParGridFunction::ProjectDiscCoefficient(VectorCoefficient &vcoeff,
                                             AvgType type)```  and more particularly within ```GridFunction::AccumulateAndCountZones``` at line 1259 when ```zones_per_vdof = 0;``` is called.

**Comment by v-dobrev** [comment]
Hmm, I looked at the code and I don't understand how `zones_per_vdof = 0;` can cause a problem -- this array is never copied/moved to GPU as far as I can tell. What is the error message that you get?

**Comment by rcarson3** [comment]
It looks like I was referencing the wrong line earlier sorry about that. I re-ran it through gdb and am now seeing the invalid host pointer assert being thrown in the 

```c++
(*this)(ldof) += isign*vals[j];
```
at[ line 1391](https://github.com/mfem/mfem/blob/master/fem/gridfunc.cpp#L1391)

Also, we can move this discussion to an issue, since the error doesn't appear to be located in the code modified by this PR.

**Comment by pazner** [comment]
> Although, it might be nice to note somewhere that making use of the compressed format is recommended if data space savings are desired. Since, I found that the binary could sometimes result in about 2x as much data as compared to the regular text for some of my simulations.

Yeah, that's a good point. Compression is turned on by default as long as MFEM is linked with zlib. For uncompressed data you could consider using `BINARY32` to reduce the output file size. Of course the comparison to ASCII output will depend on how many digits of precision you output.

**Pull request merged**